### DISCUSSION TITLE
Parquet V3 predicate pushdown to  S3/Azure object storage ?

### DISCUSSION THREAD

**Discussion by ovidiubuligan** [grafana/tempo#3047] [General]
Both S3 and azure blob storage support a features like `Predicate Pushdown` And `Projection pushdown` which drastically reduces the ammount of data to be read from parquet files.  http://peter-hoffmann.com/2020/understand-predicate-pushdown-on-rowgroup-level-in-parquet-with-pyarrow-and-python.html  .

Are these predicate pushdowns considered ? How hard would It be to implement in tempo ?

**Comment by joe-elliott** [comment]
So we do implement predicate pushdown but it is unrelated to our backend storage and works in all versions of Parquet.

The engine makes a fetch request here:

https://github.com/grafana/tempo/blob/0542bc304a570a44c1ad9c638a18c83d3ac3be3a/pkg/traceql/engine.go#L37

Each element of the AST is responsible for building its own conditions. Here is where the binary operator will create conditions and push them down to the fetch layer:

https://github.com/grafana/tempo/blob/0542bc304a570a44c1ad9c638a18c83d3ac3be3a/pkg/traceql/ast_conditions.go#L34

In the fetch layer if an operation is present we will build an iterator that asserts the operation as close to the data as possible. If no operation is present we return the data all the way back to the engine and let it decide how to handle the situation.

https://github.com/grafana/tempo/blob/0542bc304a570a44c1ad9c638a18c83d3ac3be3a/tempodb/encoding/vparquet3/block_traceql.go#L1478

**Comment by ovidiubuligan** [comment]
Thank you for the insight  , I actually wanted to link to an article about `S3 Select` that can actually do the filtering at the S3 level. 
https://docs.aws.amazon.com/AmazonS3/latest/userguide/selecting-content-from-objects.html 


Searching further It may not necessarily always be better as the issue opened In ClickHouse more than 2 years ago with no interest: 
https://github.com/ClickHouse/ClickHouse/issues/19275 

You can close the discussion.
